sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/_shared/exceptions.py

```diff
@@ -0,0 +1,54 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import functools
+from typing import TYPE_CHECKING
+
+import six
+
+from azure.core.exceptions import ResourceExistsError, ResourceNotFoundError
+
+from ._generated.v2016_10_01.models import KeyVaultErrorException as KeyVaultError2016
+from ._generated.v7_0.models import KeyVaultErrorException as KeyVaultErrorV7
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import
+    from typing import Union
+
+    KeyVaultErrorExceptionType = Union[KeyVaultError2016, KeyVaultErrorV7]
+
+_error_map = {404: ResourceNotFoundError, 409: ResourceExistsError}
+
+
+def raise_from_key_vault_exception(exception):
+    # type: (KeyVaultErrorExceptionType) -> None
+    """Raises more specific exception types from autorest-generated KeyVaultErrorException"""
+
+    new_class = None
+    try:
+        new_class = _error_map[exception.response.status_code]
+    except (AttributeError, KeyError):
+        # exception has no response/status code, or status code isn't in the error map
+        pass
+
+    if new_class:
+        new_exception = new_class(message=exception.message, response=exception.response)
+        try:
+            new_exception.error = exception.error
+        except AttributeError:
+            pass
+        six.raise_from(new_exception, exception)
+
+    raise exception
```

Comment:
I think this could be simplified with less levels of nesting
```
if not hasattr(exception.response, 'status_code'):
    raise exception

if exception.response.status_code not in _error_map:
    raise exception

new_exc =  _error_map[exception.response.status_code](message=exception.message, response=exception.response)
try:
    new_exc.error = exception.error
except AttributeError: 
    pass
six.raise_from(new_exc, exception)

Reply:
You're correct, I'll make it shorter and flatter. I wrote this originally to handle exceptions which may not have a `response`, but that's excessively defensive given the scope is now restricted to `KeyVaultErrorException`, which generated clients raise only after receiving a response.

Reply:
If that's a concern
```
if not hasattr(exception, 'response'): raise exception
``` 
at the top should handle things. (I'm definitely a fan of "bailing early" to keep things flatter)